src/typing/env_js.ml

```diff
@@ -75,33 +80,58 @@ let find_env ?(for_type=false) cx x reason =
   let rec loop = function
     | [] -> find_global cx x reason
     | scope::scopes ->
-        (match SMap.get x !scope with
+        let entries = scope.entries in
+        (match SMap.get x !entries with
         | Some entry when (not entry.for_type || for_type) -> scope
         | _ -> loop scopes)
   in
   loop (!env)
 
 let get_from_scope x scope =
-  SMap.find_unsafe x !scope
+  let entries = scope.entries in
+  SMap.find_unsafe x !entries
 
 let set_in_scope x entry scope =
-  scope := !scope |> SMap.add x entry
+  let entries = scope.entries in
+  entries := !entries |> SMap.add x entry
 
 let exists_in_scope x scope =
-  SMap.get x !scope
+  let entries = scope.entries in
+  SMap.get x !entries
+
+let set_in_env x entry scope_kind =
+  let rec loop = function
+    | [] -> set_in_scope x entry global_scope
+    | scope::scopes ->
+        if (scope_kind = LexicalScope || scope.kind = scope_kind) then
```

Comment:
Ah I think you're right -- I was getting `scope_kind` and `scope.kind` reversed in my head. Cool